[AOSS-1057] add modal-dialog module

mattpepper opened the pull request:
[AOSS-1057]
 - adds accessible & responsive modal-dialog module CSS and JavaScript (_modal-dialog.scss, modalDialog.js)
 - adds modal dialog module test specs and fixture
 - includes update to modules/_notifications.scss
   - adds '.highlight-message--light' variant
      - black message on white background
   - updates  '.highlight-message--dark' variant
      - adds '.highlight-message--dark-text' to enable specific styling of child elements


![Modal Dialog module](https://cloud.githubusercontent.com/assets/5468091/13431081/beba1cfc-dfbf-11e5-8d5f-b8759301ad21.gif)

![Modal Dialog module - with VoiceOver utility output](https://cloud.githubusercontent.com/assets/5468091/13431140/fad2d4e0-dfbf-11e5-8b5c-67944c7831a8.gif)

feedmypixel commented on assets/javascripts/application.js:
```diff
@@ -167,5 +168,6 @@ $(function() {
   accordion();
   charCounter();
   addRemove();
+  modalDialog().init();
```
you dont need the `.init()` here, follow the pattern shown above

mattpepper replied:
I'll remove modalDialog.init() from application.js and move the code inline at the end of the modalDialog.js

feedmypixel commented on assets/javascripts/modules/modalDialog.js:
```diff
@@ -0,0 +1,351 @@
+      setModalDialogDimensions();
+      setup();
+    }
+    else {
```
So if there is no modal on the page it will come into this part of the conditional, so thats all pages without a modal right will fire this `.hide()` method.

mattpepper replied:
hide() seems unnecessary in this conditional - i'll remove else clause if this is found to be true

feedmypixel commented on assets/javascripts/modules/modalDialog.js:
```diff
@@ -0,0 +1,351 @@
+module.exports = function() {
+  var $modalDialog = $('.modal-dialog'),
+    $container = $('.modal-dialog__inner'),
+    $contents = $('.modal-dialog__content'),
```
can this be more descriptive? `$contents` doesn't tell me much `$modalContents` maybe?

mattpepper replied:
agreed - changed to $modalContents

rpowis replied:
There's a lot of selector querying happening here. Can you cache the first and find on it?

```
var $modalDialog = $('.modal-dialog'),
    $container = $modalDialog.find('.modal-dialog__inner'),
    $contents = $modalDialog.find('.modal-dialog__content'),
```

etc.

feedmypixel commented on assets/javascripts/modules/modalDialog.js:
```diff
@@ -0,0 +1,351 @@
+    $contents.each(function() {
+      var $this = $(this);
+      $this.setHidden($this.getHidden());
+    });
```
this is more a style thing but I don't really know what this method is doing?

mattpepper replied:
added comment - this method sets the appropriate 'hidden' attributes/props value on module content elements as per the hidden attribute assigned/not assigned in the markup.

feedmypixel replied:
Maybe a preference here but can you comment through code? as in variable and function names, people tend not to read comments right?

feedmypixel commented on assets/javascripts/modules/modalDialog.js:
```diff
@@ -0,0 +1,351 @@
+  /**
+   * Set hidden properties & attributes
+   * @param {boolean} isHidden
+   */
```
FYI for `setHidden` and `getHidden` we have some work for controlling aria visibility tags in `visibility.js` https://github.com/hmrc/assets-frontend/blob/master/assets/javascripts/modules/visibility.js

mattpepper replied:
I'll take a look.

feedmypixel commented on assets/javascripts/modules/modalDialog.js:
```diff
@@ -0,0 +1,351 @@
+   * Get hidden property or attribute
+   * @returns {boolean|string}
+   */
+  $.fn.getHidden = function() {
```
would these functions `setHidden` and `getHidden` (if needed, see above commet) be useful outside of this module?

mattpepper replied:
possibly - i'll have a look at visibility.js

feedmypixel commented on assets/javascripts/modules/modalDialog.js:
```diff
@@ -0,0 +1,351 @@
+  /**
+   * Fire Google Analytics event tracking
+   */
+  function fireEventTracking(gaEvent) {
```
feels like this googleTracking could be in a separate module?

mattpepper replied:
That would be 'stageprompt', which this requires and uses

feedmypixel commented on assets/scss/modules/_modal-dialog.scss:
```diff
@@ -0,0 +1,141 @@
+$modal-dialog-background: rgba(255, 255, 255, 1);
+$modal-dialog-border: em(5) solid rgba(0, 0, 0, 1);
+$modal-dialog-box-shadow: 0 em(2) 0 #6f777b;
+$modal-dialog-yellow: $attorney-yellow; //#ffbf47;
```
these should be in `base/_colours.scss` right?

mattpepper replied:
$mask-opacity certainly

The others i'll look at renaming and moving

rpowis replied:
Try to use the existing palette if there's a close enough alternative too

feedmypixel commented on assets/scss/modules/_modal-dialog.scss:
```diff
@@ -0,0 +1,141 @@
+// $px-value - integer: size of square clip dimension.
+//
+// Styleguide: sass.mixins.set-square-clip
+@mixin set-square-clip($for-ie: false, $px-value: 0) {
```
this can go in `_mixins.scss`

mattpepper replied:
OK, will move

feedmypixel commented on assets/scss/modules/_modal-dialog.scss:
```diff
@@ -0,0 +1,141 @@
+    padding: 0;
+    position: absolute;
+    width: 1px;
+    z-index: 0;
```
IE stuff should be in the IE stylesheet really or use the IE mixin, I'm not sure whats preferred here

feedmypixel commented on assets/test/specs/modalDialog.spec.js:
```diff
@@ -0,0 +1,233 @@
+    it('when user uses escape key', function() {
+      $documentOpenLink = $actions.find('[data-modaldialog-action="open"][data-modaldialog-target="content-1"]');
+
+      //var e = $.Event("keyup");
```
please remove commented out code

mattpepper replied:
done

feedmypixel commented on assets/test/specs/modalDialog.spec.js:
```diff
@@ -0,0 +1,233 @@
+        $actions.find('[data-modaldialog-action="open"][data-modaldialog-target="content-1"]').click();
+      });
+
+      it('content is visible', function() {
```
it has been mentioned previously when I did some tests ideally to use the word *should* in tests

mattpepper replied:
done

feedmypixel commented on assets/javascripts/modules/modalDialog.js:
```diff
@@ -0,0 +1,351 @@
+   * Set the dimensions of the modal dialog's inner container to fit the view-port.
+   * @param {HTMLElement} $element - content element
+   */
+  function setDimensions($element) {
```
I do wonder if this can be done with `css` ?

feedmypixel commented:
@mattpepper having a look through my first thoughts are there is quite a lot of code here. It feels like some can be split out into other files, potentially some JavaScript work could be being done in css and there are some global methods that may be of use to other JavaScript work, although questionable if it is right to place them on the jQuery object as functions.
I know @rpowis was concerned fundamentally with the use of modal diolog windows and if this task could be approached in a different way.

It would be good to sit down with you and walk through this code. Maybe sometime next week?

rpowis commented:
I'm still concerned about the use of modals, but more from the point of it being a lazy design solution for the particular use case @mattpepper described to me. I still don't see why `details/summary` wouldn't do the job.

It's apparently been discussed among the design community and tested though so it might be easier to add it but flagged as `For review` or `Experimental`. Then raise it with designers to get it removed rather than hold everything up. The worry of course is that if it's in there it'll get used by others.

rpowis commented on assets/javascripts/modules/modalDialog.js:
```diff
@@ -0,0 +1,351 @@
+    $contents = $('.modal-dialog__content'),
+    contentShowing = '.modal-dialog__content:not([hidden])',
+    $doc = $(document),
+    $body = $('body'),
```
You don't need to use jQuery's selector engine to get `$body` and `$doc` here...

`$doc = document` and `$body = document.body`

mattpepper replied:
not sure I agree - i'm calling jQuery methods / events on them that aren't available to the native objects. are you saying I shouldn't?

mattpepper commented:
I've added an "Experimental" warning to the component library entry, as per Rory's comment:

![aoss-1057-modal-component-page](https://cloud.githubusercontent.com/assets/5468091/13509854/a2a16904-e185-11e5-9892-de70d30e3a85.png)

feedmypixel commented:
@mattpepper good to see the component library stuff. How did you add the experimental flag? Did you know kss-node has one built in? 
"Experimental"
I have a task to add the JavaScript to the component library will get that done on monday. 
I think I'm also going to make the code font smaller so we can fit more in.

mattpepper commented:
@feedmypixel the markdown allows for html, so I dropped a table in with some inline styling : )

Smaller font would be good. Also, some normal md styling, like block quote, isn't working.

I addressed almost all the comments, but did not use the visibility.js

feedmypixel commented:
@mattpepper ah! Just looked and saw this. We can do this in a different way with a CSS selector in the template so it is available for all. Nice styling but we shouldn't have to put a table in for all experimental work.
Let's catch up Monday when I'm in?

feedmypixel commented:
@mattpepper please raise any styling issues on the component-library-template repo. Or raise a PR on it if you feel confident, all details on README's and wiki.

feedmypixel commented:
@mattpepper really good to see how you have gone to town on this. With a few tweaks we have an experimental style. Thanks.

feedmypixel commented:
@mattpepper the component library now visually supports the flag `Experimental:` please run `npm install hmrc/component-library-template` and ass `Experimental:` at the start of the description.

feedmypixel commented:
@mattpepper - thanks for the work on this. Lets revisit this once all items are in the Component Library and speak with UX about the user need for modal-dialogs.

rpowis commented on assets/scss/modules/_modal-dialog.scss:
```diff
@@ -0,0 +1,144 @@
+* That need cannot be met by taking the user to a new page in the same flow that contains the crucial piece of content;
+
+
+Markup:
```
This is a massive example and very specific to HMRC. Can it be made smaller and have more generic content?
